internal/terraform/context_apply_deferred_test.go

```diff
@@ -1813,22 +2229,38 @@ func (provider *deferredActionsProvider) Provider() providers.Interface {
 			}
 
 			plannedState := req.ProposedNewState
+			if key == "deferred_resource_change" {
+				deferred = &providers.Deferred{
+					Reason: providers.DeferredReasonProviderConfigUnknown,
+				}
+			}
+
 			if plannedState.GetAttr("output").IsNull() {
 				plannedStateValues := req.ProposedNewState.AsValueMap()
 				plannedStateValues["output"] = cty.UnknownVal(cty.String)
 				plannedState = cty.ObjectVal(plannedStateValues)
+			} else if plannedState.GetAttr("output").AsString() == "mark_for_replacement" {
+				requiresReplace = append(requiresReplace, cty.GetAttrPath("name"), cty.GetAttrPath("output"))
 			}
 
 			provider.plannedChanges.Set(key, plannedState)
 			return providers.PlanResourceChangeResponse{
-				PlannedState: plannedState,
+				PlannedState:    plannedState,
+				Deferred:        deferred,
+				RequiresReplace: requiresReplace,
 			}
 		},
 		ApplyResourceChangeFn: func(req providers.ApplyResourceChangeRequest) providers.ApplyResourceChangeResponse {
-			key := req.Config.GetAttr("name").AsString()
+			var key string
+			if v := req.Config.GetAttr("name"); v.IsKnown() {
```

Comment:
Only computed attributes should be unknown while planning. If we are passing none computed attributes as unknown here then something has gone wrong. If a resource has been deferred and that's why we have unknown attributes in one of its dependencies, then that dependency should also have been deferred.

Reply:
Yeah good point, I took plan as an example where from my understanding the input being the prior value can be unknown. here it makes no sense to check.